### PR TITLE
[Refactor] Extract DisaggregatedResult from MetricFrame

### PR DESCRIPTION
**Pull request by riedgar-ms**
## Description

The `MetricFrame` class has long had a smaller class within it attempting to escape. This long desired goal is achieved herein.

We introduce a `DisaggregatedResult` class which provides the core functionality of `MetricFrame` but in a more constrained form. Key points:

- The functions must be supplied in the form of a dictionary of `AnnotatedMetricFunction`s
- The data must all be in a `DataFrame` which contains _everything_ - the actual data, the sensitive features, the control features, and any sample parameters

The return values are also lighter on the Pythonic niceites, skipping the handling of the 'user supplied a single metric function' case.

Essentially, this turns `MetricFrame` into a wrapper for `DisaggregatedResult`, handling a wider variety of inputs from the user, and also 'cleaning up' some return values. In due course, I anticipate that this extraction will simplify the addition of bootstrapping to `MetricFrame`.

This is not intended to be user-facing (at least not at the current time), so documentation is light.

## Tests

- [ ] no new tests required
- [ ] new tests added
- [ ] existing tests adjusted

## Documentation

- [x] no documentation changes needed
- [ ] user guide added or updated
- [ ] API docs added or updated
- [ ] example notebook added or updated

### CODE REVIEW

**Comment by riedgar-ms** [comment]
My basic idea is that for bootstrapping, the `MetricFrame` will store an internal list of `DisaggregatedResult`s constructed from the bootstrapped evaluations, and then compute the required quantities on-demand.

**Comment by riedgar-ms** [comment]
@romanlutz @adrinjalali @MiroDudik ?

**Comment by riedgar-ms** [comment]
Ping @fairlearn/fairlearn-maintainers ?

**Review by adrinjalali — COMMENTED** [review]
Seems like this is not changing public API, so easy, but we should better document the new private API before merging.

**Comment by riedgar-ms** [comment]
> Seems like this is not changing public API, so easy, but we should better document the new private API before merging.

@adrinjalali I have improved the commenting :-)

**Comment by riedgar-ms** [comment]
@adrinjalali ? @MiroDudik ?

File: docs/user_guide/assessment.rst [hunk 1]
```diff
@@ -237,17 +237,17 @@ values for the groups:
...
     >>> mf.difference()
-    tpr       0.2
-    fpr       1.0
+    tpr      0.20
+    fpr      1.00
     sel      0.25
-    count     4.0
-    dtype: object
+    count    4.00
+    dtype: float64
```
Comment (adrinjalali) [inline docs/user_guide/assessment.rst:244]: so we ARE changing things, but I think I like these changes.

File: fairlearn/metrics/_disaggregated_metric.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+
+    This holds the internal result from a disaggregated metric
+    computation, and provides `apply_grouping()` (to cover min
+    and max), `difference()` and `ratio()` methods.
+
+    The main difference to the results computed by MetricFrame
+    is that no account is made of whether the user supplied
+    a bare function or a dictionary. Hence the results are
+    always Series or DataFrame
+    """
```
Comment (adrinjalali) [inline fairlearn/metrics/_disaggregated_metric.py:68]: would be nice to have the `Parameters` section here for the inputs and hat they are supposed to be.

File: fairlearn/metrics/_disaggregated_metric.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        """Return the metrics by group."""
+        return self._by_group
+
+    def apply_grouping(
+        self,
+        grouping_function: str,
+        control_feature_names: Optional[List[str]],
+        errors: str = "raise",
+    ) -> Union[pd.Series, pd.DataFrame]:
+        """Compute mins or maxes."""
```
Comment (adrinjalali) [inline fairlearn/metrics/_disaggregated_metric.py:95]: explain the input and output here as well. Same for the methods bellow.

**Review by adrinjalali — COMMENTED** [review]
Seems like we've got some validation we're not testing mostly.

I like the refactoring otherwise.

File: docs/user_guide/assessment.rst (hunk 1 above)
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:244]: I'm not sure why that ended up changing. It was a slight surprise.

Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:244]: Given Python's approach to typing, I was actually treating that as not-a-change :-)

File: fairlearn/metrics/_disaggregated_metric.py (hunk 2 above)
Comment (riedgar-ms) [reply fairlearn/metrics/_disaggregated_metric.py:77]: Done

File: fairlearn/metrics/_disaggregated_metric.py (hunk 3 above)
Comment (riedgar-ms) [reply fairlearn/metrics/_disaggregated_metric.py]: Done

File: docs/user_guide/assessment.rst (hunk 1 above)
Comment (riedgar-ms) [reply docs/user_guide/assessment.rst:244]: And seeing other little doc issues pop up, this might even be due to a new Pandas release. I wish I knew :-/

**Comment by riedgar-ms** [comment]
OK.... docs updated and build is passing again...

**Comment by riedgar-ms** [comment]
@adrinjalali I have fixed the code coverage holes :-)

**Review by adrinjalali — APPROVED** [review]
Thanks @riedgar-ms

**Comment by adrinjalali** [comment]
I'm happy for this to be merged, not sure if we @MiroDudik wants to have a look or not.

**Pull request merged**